### PR TITLE
Test suite

### PR DESCRIPTION
**Pull request by gkellogg**
* Initial tests.
* Reference tests from spec.
* Build HTML test manifests with cross-reference back to the spec.
* GH Action to build HTML versions of test manifests and push back to branch. (removed for now).

For #20.

***
<a href="https://pr-preview.s3.amazonaws.com/json-ld/yaml-ld/pull/86.html" title="Last updated on Sep 30, 2022, 5:20 PM UTC (d50b7a6)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/json-ld/yaml-ld/86/1976cf3...d50b7a6.html" title="Last updated on Sep 30, 2022, 5:20 PM UTC (d50b7a6)">Diff</a>

### CODE REVIEW

**Comment by gkellogg** [comment]
Despite the successful run, the action which tries to push up the built manifest files is not working. Need more time to work on this, and if anyone has more experience doing this kind of thing, please jump right in.

File: tests/README.md [hunk 1]
```diff
@@ -60,6 +59,10 @@ Implementations create their own infrastructure for running the test suite. In p
 * When comparing documents after flattening, framing or generating RDF, blank node identifiers may not be predictable. Implementations should take this into consideration. (One way to do this may be to reduce both results and _expected_ to datasets to extract a bijective mapping of blank node labels between the two datasets as described in [RDF Dataset Isomorphism](https://www.w3.org/TR/rdf11-concepts/#dfn-dataset-isomorphism)).
 * Some tests may have a `requires` property, indicating some optional behavior described by a test vocabulary term.
 
+# Building HTML Manifests
+
+The HTML versions of the test manifests are built using the Node action `npm run generate`, which also cross indexes tests with references from the specification.
```
Comment (TallTed) [inline tests/README.md:64]:
```suggestion
The HTML versions of the test manifests are built using the Node action `npm run generate`, which also cross-indexes tests with references from the specification.
```

File: tests/README.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+Tests driven from a top-level [manifest](manifest.jsonld) and are defined into [basic](basic-manifest.jsonld) and [extended](extended-manifest.jsonld) sections:
+
+* [basic](basic-manifest.jsonld) tests have _input_, _expected_ and may have _context_ or _frame_ documents.
+* [extended](extended-manifest.jsonld) tests have _input_, _expected_ and may have _context_ or _frame_ documents. Additionally, the `extendedYAML` flag is added to API calls.
+
+Tests may have a `expandContext` option, which is treated
+  as an IRI relative to the manifest.
+
+The _expected_ results with file extension ".jsonld" or ".yamlld" can be compared using [YAML-LD object comparison](#yaml-ld-object-comparison) with the processor output. Additionally, if the `ordered` option is not set, result should be expanded and compared with the expanded _expected_ document also using [YAML-LD object comparison](#yaml-ld-object-comparison).
```
Comment (TallTed) [inline tests/README.md:20]:
```suggestion
The _expected_ results with file extension `.jsonld` or `.yamlld` can be compared using [YAML-LD object comparison](#yaml-ld-object-comparison) with the processor output. Additionally, if the `ordered` option is not set, the result should be expanded and compared with the expanded _expected_ document also using [YAML-LD object comparison](#yaml-ld-object-comparison).
```

File: tests/README.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+Test results that include a context input presume that the context is provided locally, and not from the referenced location, thus the results will include the content of the context file, rather than a reference.
+
+For tests of type `jld:ToRDFTest` have an _expected_ file extension of `.nq` The results can be compared using [RDF Dataset Isomorphism](https://www.w3.org/TR/rdf11-concepts/#dfn-dataset-isomorphism).
```
Comment (TallTed) [inline tests/README.md:24]:
```suggestion
Tests of type `jld:ToRDFTest` have an _expected_ file extension of `.nq`. Their results can be compared using [RDF Dataset Isomorphism](https://www.w3.org/TR/rdf11-concepts/#dfn-dataset-isomorphism).
```

File: tests/README.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+A **NegativeEvaluationTests** looks for a string value in _expectedErrorCode_, representing the error code resulting from running the test.
+
+## YAML-LD object comparison
+
+**TODO**: Similar to JSON-LD object comparison, but uses the (Expanded) Internal Representation.
+
+If algorithms are invoked with the `ordered` flag set to `true`, simple JSON Object comparison may be used, as the order of all arrays will be preserved (except for _fromRdf_, unless the input quads are also ordered). If `ordered` is `false`, then the following algorithm will ensure arrays other than values of `@list` are compared without regard to order.
+
+YAML-LD object comparison compares JSON objects, arrays, and values recursively for equality.
```
Comment (TallTed) [inline tests/README.md:36]:
```suggestion
YAML-LD object comparison recursively compares JSON objects, arrays, and values for equality.
```

File: tests/README.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+* JSON objects are compared member by member without regard to the ordering of members within the object. Each member must have a corresponding member in the object being compared to. Values are compared recursively.
+* JSON arrays are generally compared without regard to order (the lone exception being if the referencing key is `@list`). Each item within the array must be equivalent to an item in the array being compared to by using the comparison algorithm recursively. For values of `@list`, the order of these items is significant.
```
Comment (TallTed) [inline tests/README.md:39]:
```suggestion
* JSON arrays are generally compared without regard to order (the lone exception being if the referencing key is `@list`). Each item within the array must be equivalent to an item in the array being compared to by recursively using the comparison algorithm. For values of `@list`, the order of these items is significant.
```

File: tests/README.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+* JSON values are compared using strict equality.
+* Values of `@language`, and other places where language tags may be used are specified in lowercase in the test results. Implementations should either normalize language tags for testing purposes, or compare language tags in a case-independent way.
```
Comment (TallTed) [inline tests/README.md:41]:
```suggestion
* Values of `@language`, and other places where language tags may be used, are specified in lowercase in the test results. Implementations should either normalize language tags for testing purposes, or compare language tags in a case-independent way.
```

File: tests/README.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+* Some algorithms, particularly _fromRdf_, may not preserve the order of statements listed in the input document, and provision should be taken for performing unordered array comparison, for arrays other than values of `@list`. (This may be difficult for compacted results, where array value ordering is dependent on the associated term definition).
+* Some _toRdf_ tests require the use of [JSON Canonicalization Scheme](https://tools.ietf.org/html/draft-rundgren-json-canonicalization-scheme-05) to properly generate RDF Literals from JSON literal values. This algorithm is to be used to properly compare results using [RDF Dataset Isomorphism](https://www.w3.org/TR/rdf11-concepts/#dfn-dataset-isomorphism). These tests are marked using the `useJCS` option.
+* Some _toRdf_ tests may generate results as [Generalized RDF](https://www.w3.org/TR/rdf11-concepts/#section-generalized-rdf), specifically having a blank node predicate. Technically, these are in an invalid N-Quads format; provisions must be taken to be able to parse and recognize quads having a blank node predicate.
```
Comment (TallTed) [inline tests/README.md:58]:
```suggestion
* Some _toRdf_ tests may generate results as [Generalized RDF](https://www.w3.org/TR/rdf11-concepts/#section-generalized-rdf), specifically having a blank node predicate. Technically, these are in an invalid N-Quads format; provision must be made to parse and recognize quads having a blank node predicate.
```

File: tests/README.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+# Building HTML Manifests
+
+The HTML versions of the test manifests are built using the Node action `npm run generate`, which also cross indexes tests with references from the specification.
+
+# Contributing
+
+If you would like to contribute a new test or a fix to an existing test,
+please follow these steps:
+
+1. Notify the JSON-LD mailing list, public-linked-json@w3.org,
```
Comment (TallTed) [inline tests/README.md:71]: Though I do wonder, why not suggest they create an issue on the repo?
```suggestion
1. Notify the JSON-LD Working Group mailing list, [public-linked-json@w3.org](mailto:public-linked-json@w3.org),
```

File: tests/basic-manifest.jsonld [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+      "@type": ["jld:PositiveEvaluationTest", "jld:ExpandTest"],
+      "name": "UTF-8 BOM",
+      "purpose": "A YAML-LD document MUST be encoded in UTF-8, to ensure interoperability with [[JSON]].",
+      "input": "basic/cr-utf8-positive-1-in.yamlld",
+      "expect": "basic/cr-utf8-positive-1-out.yamlld"
+    },
+    {
+      "@id": "#cr-utf8-negative-1",
+      "@type": ["jld:NegativeEvaluationTest", "jld:ExpandTest"],
+      "name": "UTF-6 BOM",
```
Comment (pchampin) [inline tests/basic-manifest.jsonld:20]: I guess this is meant to be "UTF-16"?

File: tests/README.md (hunk 8 above)
Comment (pchampin) [reply tests/README.md:71]: +1 to also add a reference to the github repo

**Review by pchampin — APPROVED** [review]
LGTM modulo minor remarks

File: tests/basic-manifest.jsonld (hunk 9 above)
Comment (TallTed) [reply tests/basic-manifest.jsonld:20]: See also apparently related line 136 —
```
                Test cr-utf8-negative-1 UTF-6 BOM
```
— and line 142 —
```
                  <dt>Name</dt><dd>UTF-6 BOM</dd></dt>
```
— of tests/basic-manifest.html

Comment (gkellogg) [reply tests/basic-manifest.jsonld:20]: Those last two are derived from the .jsonld version.

File: tests/README.md (hunk 8 above)
Comment (gkellogg) [reply tests/README.md]: Check updated document, also in generated tests/*manifest.html files.

**Comment by gkellogg** [comment]
A couple of more changes, including comments on generated files that changes should be made to the source.

I did try to make these files ignored and have them built by GH Action, but failed in that. I notice in the ePub tests, generated files are only made to the main branch when a PR is merged, which may ultimately be the way to go. @iherman may have some thoughts on this.

File: tests/README.md [hunk 10]
```diff
@@ -68,11 +68,12 @@ The HTML versions of the test manifests are built using the Node action `npm run
 If you would like to contribute a new test or a fix to an existing test,
 please follow these steps:
 
-1. Notify the JSON-LD Community Group mailing list, [public-linked-json@w3.org](mailto:public-linked-json@w3.org),
-   that you will be creating a new test or fix and the purpose of the
-   change, or [create an issue](https://github.com/json-ld/yaml-ld/issues).
+1. Notify the JSON-LD Community Group that you will be creating
+  a new test or fix and the purpose of the change
+  by [creating an issue](https://github.com/json-ld/yaml-ld/issues) on the [GitHub Repository](https://github.com/json-ld/yaml-ld),
+  or send a message to the group mailing list, [public-linked-json@w3.org](mailto:public-linked-json@w3.org).
```
Comment (TallTed) [inline tests/README.md:74]:
```suggestion
  or by sending a message to the group mailing list, [public-linked-json@w3.org](mailto:public-linked-json@w3.org).
```

**Comment by gkellogg** [comment]
Thanks for the improvements, I think it will be more useful for people once this is merged. I'm working on additional tests that will come in another PR.###

**Pull request merged**